invenio/modules/workflows/client.py

```diff
@@ -39,15 +44,25 @@ def run_workflow(wfe, data, stop_on_halt=False,
     :param wfe: workflow engine in charge of workflow execution
     """
     while True:
+        # Make a backup of the db session and replace the global one with a
+        # new clean version. This is required because the workflow payload
+        # might do horrible things with it and we need to for querying+storing
+        # the workflow state/objects.
+        session_backup = db.session
+        db.session = db.create_scoped_session()
```

Comment:
You need another connection from the pool for the payload, during the payload execution. See comment on the celery solution for a cleaner future.